File: Exp2_Hello World/MainActivity.java
package com.example.helloandroidapplication

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.graphics.Color
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val helloText = findViewById<TextView>(R.id.helloText)

        // Animate between multiple colors (like CSS keyframes)
        val colors = arrayOf(
            Color.RED,
            Color.MAGENTA,
            Color.BLUE,
            Color.CYAN,
            Color.GREEN,
            Color.YELLOW,
            Color.RED

        )

        val animator = ValueAnimator.ofFloat(0f, (colors.size - 1).toFloat())
        animator.duration = 4000
        animator.repeatCount = ValueAnimator.INFINITE
        animator.addUpdateListener { animation ->
            val position = animation.animatedValue as Float
            val index = position.toInt()
            val fraction = position - index

            val color = ArgbEvaluator().evaluate(
                fraction,
                colors[index],
                colors[(index + 1) % colors.size]
            ) as Int

            helloText.setTextColor(color)
        }

        animator.start()
    }
}

File: Exp8_TallyCounter/MainActivity.kt
package com.example.tallycounter

import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    private lateinit var counterText: TextView
    private lateinit var incrementButton: Button
    private lateinit var decrementButton: Button
    private lateinit var resetButton: Button
    private var counter = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Initialize views
        counterText = findViewById(R.id.counterText)
        incrementButton = findViewById(R.id.incrementButton)
        decrementButton = findViewById(R.id.decrementButton)
        resetButton = findViewById(R.id.resetButton)

        // Button Click Listeners
        incrementButton.setOnClickListener {
            counter++
            updateCounter()
            showToast("Nice! Count is now $counter 📈")
        }

        decrementButton.setOnClickListener {
            counter--
            updateCounter()
            showToast("Oops! Count dropped to $counter 📉")
        }

        resetButton.setOnClickListener {
            counter = 0
            updateCounter()
            showToast("Counter reset. Back to zero! 🔄")
        }

        // Show initial counter
        updateCounter()
    }

    private fun updateCounter() {
        counterText.text = counter.toString()

        // Change text color based on value
        counterText.setTextColor(
            when {
                counter > 0 -> Color.parseColor("#388E3C") // Green
                counter < 0 -> Color.parseColor("#D32F2F") // Red
                else -> Color.parseColor("#1976D2")       // Blue
            }
        )
    }

    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    // Save state on rotate/background
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("counter_value", counter)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        counter = savedInstanceState.getInt("counter_value")
        updateCounter()
    }
}
